compiler/rustc_macros/src/diagnostics/diagnostic_builder.rs

```diff
@@ -182,59 +193,116 @@ impl DiagnosticDeriveVariantBuilder {
         let name = attr.path().segments.last().unwrap().ident.to_string();
         let name = name.as_str();
 
-        let mut first = true;
+        let mut set_label = false;
+        let keys = vec!["note", "help", "warning", "suggestion"];
 
         if name == "diag" {
```

Comment:
Would it be simpler to introduce a mutually exclusive `diag_new` and re-use some logic, what would parse as `slug` becomes `message` (or `label` as you have it now)?